File: tools/theme-contrast/png.mjs
/**
 * Minimaler PNG-Dekoder (nur was Chromes `--screenshot` erzeugt).
 * ───────────────────────────────────────────────────────────────────────────
 * Unterstützt: Bittiefe 8, Farbtyp 2 (RGB) und 6 (RGBA), nicht interlaced,
 * beliebig viele IDAT-Chunks. Mehr braucht ein Chrome-Screenshot nicht — und
 * mehr zu können hieße, Code zu pflegen, den nie jemand ausführt.
 *
 * Warum kein npm-Paket: das Harness soll in einem frischen Worktree ohne
 * `npm install` laufen (und ohne dass irgendwer einer Kette von Transitiv-
 * Abhängigkeiten vertrauen muss, nur um Pixel zu zählen). zlib ist in Node.
 */

import { inflateSync } from 'node:zlib';

/** @returns {{width:number,height:number,channels:number,data:Uint8Array}} */
export function decodePng(buf) {
  if (buf.readUInt32BE(0) !== 0x89504e47) throw new Error('kein PNG');

  let pos = 8;
  let width = 0;
  let height = 0;
  let channels = 0;
  const idat = [];

  while (pos < buf.length) {
    const len = buf.readUInt32BE(pos);
    const type = buf.toString('ascii', pos + 4, pos + 8);
    const body = buf.subarray(pos + 8, pos + 8 + len);
    pos += 12 + len; // len + type + data + crc

    if (type === 'IHDR') {
      width = body.readUInt32BE(0);
      height = body.readUInt32BE(4);
      const depth = body[8];
      const colorType = body[9];
      const interlace = body[12];
      if (depth !== 8) throw new Error(`Bittiefe ${depth} nicht unterstützt`);
      if (interlace !== 0) throw new Error('interlaced nicht unterstützt');
      if (colorType === 2) channels = 3;
      else if (colorType === 6) channels = 4;
      else throw new Error(`Farbtyp ${colorType} nicht unterstützt`);
    } else if (type === 'IDAT') {
      idat.push(body);
    } else if (type === 'IEND') {
      break;
    }
  }

  const raw = inflateSync(Buffer.concat(idat));
  const stride = width * channels;
  const out = new Uint8Array(width * height * channels);

  // PNG-Filter rückgängig machen (Spec 9.2). `prev` ist die bereits
  // entfilterte Vorzeile — deshalb läuft das strikt von oben nach unten.
  for (let y = 0; y < height; y++) {
    const filter = raw[y * (stride + 1)];
    const line = raw.subarray(y * (stride + 1) + 1, y * (stride + 1) + 1 + stride);
    const cur = out.subarray(y * stride, y * stride + stride);
    const prev = y > 0 ? out.subarray((y - 1) * stride, (y - 1) * stride + stride) : null;

    for (let i = 0; i < stride; i++) {
      const a = i >= channels ? cur[i - channels] : 0; // links
      const b = prev ? prev[i] : 0; // oben
      const c = prev && i >= channels ? prev[i - channels] : 0; // oben-links
      let v = line[i];
      switch (filter) {
        case 0:
          break;
        case 1:
          v += a;
          break;
        case 2:
          v += b;
          break;
        case 3:
          v += (a + b) >> 1;
          break;
        case 4: {
          const p = a + b - c;
          const pa = Math.abs(p - a);
          const pb = Math.abs(p - b);
          const pc = Math.abs(p - c);
          v += pa <= pb && pa <= pc ? a : pb <= pc ? b : c;
          break;
        }
        default:
          throw new Error(`Filter ${filter} unbekannt`);
      }
      cur[i] = v & 0xff;
    }
  }

  return { width, height, channels, data: out };
}

/** Pixel als sRGB 0..1 (Alpha ignoriert — ein Screenshot ist deckend). */
export function pixel(img, x, y) {
  const i = (y * img.width + x) * img.channels;
  return [img.data[i] / 255, img.data[i + 1] / 255, img.data[i + 2] / 255];
}

File: tools/theme-contrast/png.py
"""Minimal dependency-free PNG reader (8-bit RGB/RGBA, non-interlaced).

Chrome's `--screenshot` writes exactly that flavour, and this box has neither
Pillow nor numpy, so the twenty lines of un-filtering live here instead of in a
wheel. Returns (width, height, rowbytes) with 3 or 4 channels per pixel.
"""

import struct
import zlib


def read_png(path):
    with open(path, 'rb') as fh:
        data = fh.read()
    if data[:8] != b'\x89PNG\r\n\x1a\n':
        raise ValueError('not a PNG: %s' % path)

    pos = 8
    width = height = depth = ctype = None
    idat = []
    while pos < len(data):
        (length,) = struct.unpack('>I', data[pos:pos + 4])
        ctag = data[pos + 4:pos + 8]
        body = data[pos + 8:pos + 8 + length]
        pos += 12 + length  # length + tag + body + crc
        if ctag == b'IHDR':
            width, height, depth, ctype, _comp, _filt, interlace = struct.unpack('>IIBBBBB', body)
            if depth != 8 or ctype not in (2, 6) or interlace != 0:
                raise ValueError('unsupported PNG (depth=%s ctype=%s interlace=%s)'
                                 % (depth, ctype, interlace))
        elif ctag == b'IDAT':
            idat.append(body)
        elif ctag == b'IEND':
            break

    nch = 3 if ctype == 2 else 4
    raw = zlib.decompress(b''.join(idat))
    stride = width * nch
    out = bytearray(height * stride)
    prev = bytearray(stride)
    rp = 0
    for y in range(height):
        ftype = raw[rp]
        rp += 1
        line = bytearray(raw[rp:rp + stride])
        rp += stride
        if ftype == 1:  # Sub
            for i in range(nch, stride):
                line[i] = (line[i] + line[i - nch]) & 0xFF
        elif ftype == 2:  # Up
            for i in range(stride):
                line[i] = (line[i] + prev[i]) & 0xFF
        elif ftype == 3:  # Average
            for i in range(stride):
                left = line[i - nch] if i >= nch else 0
                line[i] = (line[i] + ((left + prev[i]) >> 1)) & 0xFF
        elif ftype == 4:  # Paeth
            for i in range(stride):
                a = line[i - nch] if i >= nch else 0
                b = prev[i]
                c = prev[i - nch] if i >= nch else 0
                p = a + b - c
                pa, pb, pc = abs(p - a), abs(p - b), abs(p - c)
                pr = a if (pa <= pb and pa <= pc) else (b if pb <= pc else c)
                line[i] = (line[i] + pr) & 0xFF
        elif ftype != 0:
            raise ValueError('bad filter %d' % ftype)
        out[y * stride:(y + 1) * stride] = line
        prev = line
    return width, height, nch, bytes(out)


_LIN = [((c / 255.0) / 12.92) if (c / 255.0) <= 0.04045
        else (((c / 255.0) + 0.055) / 1.055) ** 2.4 for c in range(256)]


def luminance(r, g, b):
    """WCAG 2.1 relative luminance from 8-bit sRGB."""
    return 0.2126 * _LIN[r] + 0.7152 * _LIN[g] + 0.0722 * _LIN[b]


def contrast(l1, l2):
    hi, lo = (l1, l2) if l1 >= l2 else (l2, l1)
    return (hi + 0.05) / (lo + 0.05)
